hs_composite_resource/models.py

```diff
@@ -111,14 +179,31 @@ def can_set_folder_to_fileset(self, dir_path):
         :param dir_path: Resource file directory path (full folder path starting with resource id)
         for which the FileSet aggregation to be set
 
-        :return If the specified folder is already represents an aggregation or does
-        not contain any files then returns False, otherwise True
+        :return
+        If the specified folder is already represents an aggregation, return False
+        if the specified folder does not contain any files, return False
+        if any of the parent folders is a model program aggregation, return False
+        if any of the parent folders is a model instance aggregation, return False
+        otherwise, return True
+
+        Note: A fileset aggregation is not allowed inside a model program or model instance aggregation. One
+        fileset aggregation can contain any other aggregation types including fileset aggregation
         """
 
         if self.get_folder_aggregation_object(dir_path) is not None:
             # target folder is already an aggregation
             return False
 
+        # checking all parent folders
+        path = os.path.dirname(dir_path)
+        while '/' in path:
+            parent_aggr = self.get_folder_aggregation_object(path)
+            if parent_aggr is not None and (parent_aggr.is_model_program or parent_aggr.is_model_instance):
```

Comment:
This is another rule that is in the composite resource that really belongs in the logical files

Reply:
Moved to logical file class